File: workersc++/WorkerBase.cpp
// Copyright (C) 2024 INAF
// This software is distributed under the terms of the BSD-3-Clause license
//
// Authors:
//
//    Andrea Bulgarelli <andrea.bulgarelli@inaf.it>
//

#include <string>
#include <iostream>
#include <json/json.h> // JSON library
#include <zmq.hpp>     // ZeroMQ library
#include "Logger.h"    // Assume you have a Logger class similar to Python's Logger
#include "Supervisor.h"
#include "WorkerManager.h"

// Base class for workers
class WorkerBase {
public:
    WorkerBase() = default;

    // Initialize the worker with manager, supervisor, and names
    void init(WorkerManager* manager, Supervisor* supervisor, const std::string& workersname, const std::string& fullname) {
        this->manager = manager;
        this->supervisor = supervisor;
        this->logger = supervisor->getLogger(); // Assuming Supervisor has a method getLogger()
        this->workersname = workersname;
        this->fullname = fullname;
    }

    // To be reimplemented in derived classes
    virtual void config(const Json::Value& configuration) {
        // Extract the pidtarget
        std::string pidtarget = configuration["header"]["pidtarget"].asString();

        // Check if the configuration is meant for this worker
        if (pidtarget == workersname || pidtarget == fullname) {
            std::cout << "Received config: " << configuration << std::endl;
        } else {
            return;
        }
    }

    // To be reimplemented in derived classes
    virtual void process_data(const std::string& data, int priority) {
        // Process data implementation to be provided in derived classes
    }

protected:
    WorkerManager* manager = nullptr;
    Supervisor* supervisor = nullptr;
    Logger* logger = nullptr; // Assuming Logger is a custom class similar to Python's Logger
    std::string workersname;
    std::string fullname;
};

File: workers/WorkerBase.py
# Copyright (C) 2024 INAF
# This software is distributed under the terms of the BSD-3-Clause license
#
# Authors:
#
#    Andrea Bulgarelli <andrea.bulgarelli@inaf.it>
#

import json
import zmq
from logging import Logger
from Supervisor import Supervisor
from WorkerManager import WorkerManager

class WorkerBase():
	def __init__(self):
		pass
	
	def init(self, manager: WorkerManager, supervisor: Supervisor, workersname: str, fullname: str):
		self.manager = manager
		self.supervisor = supervisor
		self.logger = supervisor.logger
		self.workersname = workersname
		self.fullname = fullname

	#to be reimplemented ####
	def config(self, conf_message):
		# Get configuration
		configuration = conf_message['config']['config']
		print(f"Received config: {configuration}")
		self.logger.system(f"Received config: {configuration}")
		return configuration


	

	#to be reimplemented ####
	def process_data(self, data, priority):
		pass

